modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,815 @@
+        # Using stack and generators to avoid the ``RecursionError``s.
+        stack.append(self._deconstruct_chain(self, output, stack, result_consumers))
+        while stack:
+            try:
+                gen = stack.pop()
+                next_gen = next(gen)
+                stack.append(gen)
+                stack.append(next_gen)
+            except StopIteration:
+                pass
+        return result_consumers, output
+
+    @classmethod
+    def _deconstruct_chain(
+        cls,
+        de: "DeferredExecution",
+        output: List,
+        stack: List,
+        result_consumers: List["DeferredExecution"],
+    ):
+        """
+        Deconstruct the specified DeferredExecution chain.
+
+        Parameters
+        ----------
+        de : DeferredExecution
+            The chain to be deconstructed.
+        output : list
+            Put the arguments to this list.
+        stack : list
+            Used to eliminate recursive calls, that may lead to the RecursionError.
+        result_consumers : list of DeferredExecution
+            The result consumers.
+
+        Yields
+        ------
+        Generator
+            The ``_deconstruct_list()`` generator.
+        """
+        out_append = output.append
+        out_extend = output.extend
+        while True:
+            de.ref_count(-1)
+            if (out_pos := getattr(de, "out_pos", None)) and not de.has_result:
+                out_append(_Tag.REF)
+                out_append(out_pos)
+                output[out_pos] = out_pos
+                if de.ref_counter == 0:
+                    output[out_pos + 1] = 0
+                    result_consumers.remove(de)
+                break
+            elif not isinstance(data := de.data, DeferredExecution):
+                if isinstance(data, ListOrTuple):
+                    yield cls._deconstruct_list(
+                        data, output, stack, result_consumers, out_append
+                    )
+                else:
+                    out_append(data)
+                if not de.has_result:
+                    stack.append(de)
+                break
+            else:
+                stack.append(de)
+                de = data
+
+        assert stack and isinstance(stack[-1], DeferredExecution)
+        while stack and isinstance(stack[-1], DeferredExecution):
+            de: DeferredExecution = stack.pop()
+            args = de.args
+            kwargs = de.kwargs
+            out_append(de.func)
+            if de.flat_args:
+                out_append(len(args))
+                out_extend(args)
+            else:
+                out_append(-1)
+                yield cls._deconstruct_list(
+                    args, output, stack, result_consumers, out_append
+                )
+            if de.flat_kwargs:
+                out_append(len(kwargs))
+                for item in kwargs.items():
+                    out_extend(item)
+            else:
+                out_append(-1)
```

Comment:
What does `-1` mean?

Reply:
It's described in the doc:
If it is -1, it means that the method arguments contain list and/or DeferredExecution (chain) objects.

Reply:
You have a separate class with constants that you use for encoding, why was this value not added?

Reply:
Because in a basic case, this field contains the list length. -1 means that the next value needs to decoded sequentially, one-by-one.